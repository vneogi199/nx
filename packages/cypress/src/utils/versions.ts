export const nxVersion = require('../../package.json').version;
export const eslintPluginCypressVersion = '^2.13.4';
export const typesNodeVersion = '16.11.7';
export const cypressViteDevServerVersion = '^2.2.1';
export const cypressVersion = '^13.0.0';
export const cypressWebpackVersion = '^2.0.0';
export const webpackHttpPluginVersion = '^5.5.0';
export const viteVersion = '~4.3.9';
export const htmlWebpackPluginVersion = '^5.5.0';
