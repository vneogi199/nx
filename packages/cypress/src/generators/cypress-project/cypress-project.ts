import {
  addDependenciesToPackageJson,
  addProjectConfiguration,
  formatFiles,
  generateFiles,
  GeneratorCallback,
  getProjects,
  joinPathFragments,
  logger,
  offsetFromRoot,
  ProjectConfiguration,
  readProjectConfiguration,
  runTasksInSerial,
  stripIndents,
  toJS,
  Tree,
  updateJson,
} from '@nx/devkit';
import { determineProjectNameAndRootOptions } from '@nx/devkit/src/generators/project-name-and-root-utils';
import { checkAndCleanWithSemver } from '@nx/devkit/src/utils/semver';
import { getRelativePathToRootTsConfig } from '@nx/js';
import { Linter } from '@nx/eslint';
import { join } from 'path';
import { major } from 'semver';
import { addLinterToCyProject } from '../../utils/add-linter';
import { installedCypressVersion } from '../../utils/cypress-version';
import { cypressVersion, viteVersion } from '../../utils/versions';
import { cypressInitGenerator } from '../init/init';
import { Schema } from './schema';

export interface CypressProjectSchema extends Schema {
  projectName: string;
  projectRoot: string;
  rootProject: boolean;
}

function createFiles(tree: Tree, options: CypressProjectSchema) {
  // if not installed or >v10 use v10 folder
  // else use v9 folder
  const cypressVersion = installedCypressVersion();
  const cypressFiles =
    cypressVersion && cypressVersion < 10 ? 'v9-and-under' : 'v10-and-after';

  generateFiles(
    tree,
    join(__dirname, './files', cypressFiles),
    options.projectRoot,
    {
      tmpl: '',
      ...options,
      project: options.project || 'Project',
      ext: options.js ? 'js' : 'ts',
      offsetFromRoot: offsetFromRoot(options.projectRoot),
      rootTsConfigPath: getRelativePathToRootTsConfig(
        tree,
        options.projectRoot
      ),
      bundler: options.bundler,
    }
  );

  if (cypressVersion && cypressVersion < 7) {
    updateJson(tree, join(options.projectRoot, 'cypress.json'), (json) => {
      json.pluginsFile = './src/plugins/index';
      return json;
    });
  } else if (cypressVersion < 10) {
    const pluginPath = join(options.projectRoot, 'src/plugins/index.js');
    if (tree.exists(pluginPath)) {
      tree.delete(pluginPath);
    }
  }

  if (options.js) {
    toJS(tree);
  }
}

function addProject(tree: Tree, options: CypressProjectSchema) {
  let e2eProjectConfig: ProjectConfiguration;

  const detectedCypressVersion =
    installedCypressVersion() ??
    major(checkAndCleanWithSemver('cypress', cypressVersion));

  const cypressConfig =
    detectedCypressVersion < 10 ? 'cypress.json' : 'cypress.config.ts';

  if (options.baseUrl) {
    e2eProjectConfig = {
      root: options.projectRoot,
      sourceRoot: joinPathFragments(options.projectRoot, 'src'),
      projectType: 'application',
      targets: {
        e2e: {
          executor: '@nx/cypress:cypress',
          options: {
            cypressConfig: joinPathFragments(
              options.projectRoot,
              cypressConfig
            ),
            baseUrl: options.baseUrl,
            testingType: 'e2e',
          },
        },
      },
      tags: [],
      implicitDependencies: options.project ? [options.project] : undefined,
    };
  } else if (options.project) {
    const project = readProjectConfiguration(tree, options.project);

    if (!project.targets) {
      logger.warn(stripIndents`
      NOTE: Project, "${options.project}", does not have any targets defined and a baseUrl was not provided. Nx will use
      "${options.project}:serve" as the devServerTarget. But you may need to define this target within the project, "${options.project}".
      `);
    }
    const devServerTarget =
      project.targets?.serve && project.targets?.serve?.defaultConfiguration
        ? `${options.project}:serve:${project.targets.serve.defaultConfiguration}`
        : `${options.project}:serve`;
    e2eProjectConfig = {
      root: options.projectRoot,
      sourceRoot: joinPathFragments(options.projectRoot, 'src'),
      projectType: 'application',
      targets: {
        e2e: {
          executor: '@nx/cypress:cypress',
          options: {
            cypressConfig: joinPathFragments(
              options.projectRoot,
              cypressConfig
            ),
            devServerTarget,
            testingType: 'e2e',
          },
          configurations: {
            production: {
              devServerTarget: `${options.project}:serve:production`,
            },
          },
        },
      },
      tags: [],
      implicitDependencies: options.project ? [options.project] : undefined,
    };
    if (project.targets?.['serve-static']) {
      e2eProjectConfig.targets.e2e.configurations.ci = {
        devServerTarget: `${options.project}:serve-static`,
      };
    }
  } else {
    throw new Error(`Either project or baseUrl should be specified.`);
  }

  if (detectedCypressVersion < 7) {
    e2eProjectConfig.targets.e2e.options.tsConfig = joinPathFragments(
      options.projectRoot,
      'tsconfig.json'
    );
  }
  addProjectConfiguration(tree, options.projectName, e2eProjectConfig);
}

/**
 * @deprecated use cypressE2EConfigurationGenerator instead
 **/
export async function cypressProjectGenerator(host: Tree, schema: Schema) {
  return await cypressProjectGeneratorInternal(host, {
    projectNameAndRootFormat: 'derived',
    ...schema,
  });
}

/**
 * @deprecated use cypressE2EConfigurationGenerator instead
 **/
export async function cypressProjectGeneratorInternal(
  host: Tree,
  schema: Schema
) {
  const options = await normalizeOptions(host, schema);
  const tasks: GeneratorCallback[] = [];
  const cypressVersion = installedCypressVersion();
  // if there is an installed cypress version, then we don't call
  // init since we want to keep the existing version that is installed
  if (!cypressVersion) {
    tasks.push(await cypressInitGenerator(host, options));
  }

  if (schema.bundler === 'vite') {
    tasks.push(
      addDependenciesToPackageJson(
        host,
        {},
        {
          vite: viteVersion,
        }
      )
    );
  }

  createFiles(host, options);
  addProject(host, options);
  const installTask = await addLinterToCyProject(host, {
    ...options,
    cypressDir: 'src',
    linter: schema.linter,
    project: options.projectName,
    overwriteExisting: true,
  });
  tasks.push(installTask);
  if (!options.skipFormat) {
    await formatFiles(host);
  }
  return runTasksInSerial(...tasks);
}

async function normalizeOptions(
  host: Tree,
  options: Schema
): Promise<CypressProjectSchema> {
  let maybeRootProject: ProjectConfiguration;
  let isRootProject = false;

  const projects = getProjects(host);
  // nx will set the project option for generators when ran within a project.
  // since the root project will always be set for standalone projects we can just check it here.
  if (options.project) {
    maybeRootProject = projects.get(options.project);
  }

  if (
    maybeRootProject?.root === '.' ||
    // should still check to see if we are in a standalone based workspace
    (!maybeRootProject &&
      Array.from(projects.values()).some((config) => config.root === '.'))
  ) {
    isRootProject = true;
  }

  let { projectName, projectRoot } = await determineProjectNameAndRootOptions(
    host,
    {
      name: options.name,
      projectType: 'application',
      directory: isRootProject ? options.name : options.directory,
      projectNameAndRootFormat: isRootProject
        ? 'as-provided'
        : options.projectNameAndRootFormat,
      callingGenerator: '@nx/cypress:cypress-project',
    }
  );

  options.linter = options.linter || Linter.EsLint;
  options.bundler = options.bundler || 'webpack';
  return {
    ...options,
    // other generators depend on the rootProject flag down stream
    rootProject: isRootProject,
    projectName,
    projectRoot,
  };
}

export default cypressProjectGenerator;
