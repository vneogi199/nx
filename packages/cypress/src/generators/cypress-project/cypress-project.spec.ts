import {
  addProjectConfiguration,
  readJson,
  readProjectConfiguration,
  Tree,
  updateProjectConfiguration,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { cypressProjectGenerator } from './cypress-project';
import { Schema } from './schema';
import { Linter } from '@nx/eslint';
import { installedCypressVersion } from '../../utils/cypress-version';
import { cypressInitGenerator } from '../init/init';

jest.mock('../../utils/cypress-version');
jest.mock('../init/init');
describe('Cypress Project', () => {
  let tree: Tree;
  const defaultOptions: Omit<Schema, 'name' | 'project'> = {
    linter: Linter.EsLint,
  };
  let mockedInstalledCypressVersion: jest.Mock<
    ReturnType<typeof installedCypressVersion>
  > = installedCypressVersion as never;
  let mockInitCypress: jest.Mock<ReturnType<typeof cypressInitGenerator>> =
    cypressInitGenerator as never;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    addProjectConfiguration(tree, 'my-app', {
      root: 'my-app',
      targets: {
        serve: {
          executor: 'serve-executor',
          options: {},
          configurations: {
            production: {},
          },
        },
      },
    });

    addProjectConfiguration(tree, 'my-dir-my-app', {
      root: 'my-dir/my-app',
      targets: {
        serve: {
          executor: 'serve-executor',
          options: {},
          configurations: {
            production: {},
          },
        },
      },
    });
  });
  afterEach(() => jest.clearAllMocks());

  it('should call init if cypress is not installed', async () => {
    mockedInstalledCypressVersion.mockReturnValue(null);

    await cypressProjectGenerator(tree, {
      ...defaultOptions,
      name: 'my-app-e2e',
      project: 'my-app',
      projectNameAndRootFormat: 'as-provided',
    });

    expect(mockInitCypress).toHaveBeenCalled();
  });

  it('should call not init if cypress is installed', async () => {
    mockedInstalledCypressVersion.mockReturnValue(10);

    await cypressProjectGenerator(tree, {
      ...defaultOptions,
      name: 'my-app-e2e',
      project: 'my-app',
      projectNameAndRootFormat: 'as-provided',
    });

    expect(mockInitCypress).not.toHaveBeenCalled();
  });

  describe('> v10', () => {
    beforeEach(() => {
      mockedInstalledCypressVersion.mockReturnValue(10);
    });

    it('should generate files for v10 and above', async () => {
      await cypressProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
        projectNameAndRootFormat: 'as-provided',
      });

      expect(tree.exists('my-app-e2e/cypress.config.ts')).toBeTruthy();
      expect(tree.exists('my-app-e2e/src/fixtures/example.json')).toBeTruthy();
      expect(tree.exists('my-app-e2e/src/e2e/app.cy.ts')).toBeTruthy();
      expect(tree.exists('my-app-e2e/src/support/app.po.ts')).toBeTruthy();
      expect(tree.exists('my-app-e2e/src/support/commands.ts')).toBeTruthy();
      expect(tree.exists('my-app-e2e/src/support/e2e.ts')).toBeTruthy();
    });

    it('should update configuration when eslint is passed', async () => {
      await cypressProjectGenerator(tree, {
        name: 'my-app-e2e',
        project: 'my-app',
        linter: Linter.EsLint,
        projectNameAndRootFormat: 'as-provided',
      });

      const project = readProjectConfiguration(tree, 'my-app-e2e');
      expect(project.targets.lint).toMatchSnapshot();
    });

    it('should not add lint target when "none" is passed', async () => {
      await cypressProjectGenerator(tree, {
        name: 'my-app-e2e',
        project: 'my-app',
        linter: Linter.None,
        projectNameAndRootFormat: 'as-provided',
      });

      const project = readProjectConfiguration(tree, 'my-app-e2e');
      expect(project.targets.lint).toBeUndefined();
    });

    it('should update tags and implicit dependencies', async () => {
      await cypressProjectGenerator(tree, {
        name: 'my-app-e2e',
        project: 'my-app',
        linter: Linter.EsLint,
        projectNameAndRootFormat: 'as-provided',
      });

      const project = readProjectConfiguration(tree, 'my-app-e2e');
      expect(project.tags).toEqual([]);
      expect(project.implicitDependencies).toEqual(['my-app']);
    });

    it('should set right path names in `cypress.config.ts`', async () => {
      await cypressProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
        projectNameAndRootFormat: 'as-provided',
      });

      const cypressConfig = tree.read('my-app-e2e/cypress.config.ts', 'utf-8');
      expect(cypressConfig).toMatchSnapshot();
    });

    it('should set right path names in `tsconfig.e2e.json`', async () => {
      await cypressProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
        projectNameAndRootFormat: 'as-provided',
      });

      const tsconfigJson = readJson(tree, 'my-app-e2e/tsconfig.json');
      expect(tsconfigJson).toMatchSnapshot();
    });

    it('should extend from tsconfig.base.json', async () => {
      await cypressProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
        projectNameAndRootFormat: 'as-provided',
      });

      const tsConfig = readJson(tree, 'my-app-e2e/tsconfig.json');
      expect(tsConfig.extends).toBe('../tsconfig.base.json');
    });

    it('should support a root tsconfig.json instead of tsconfig.base.json', async () => {
      tree.rename('tsconfig.base.json', 'tsconfig.json');

      await cypressProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
        projectNameAndRootFormat: 'as-provided',
      });

      const tsConfig = readJson(tree, 'my-app-e2e/tsconfig.json');
      expect(tsConfig.extends).toBe('../tsconfig.json');
    });

    describe('for bundler:vite', () => {
      it('should pass the bundler info to nxE2EPreset in `cypress.config.ts`', async () => {
        await cypressProjectGenerator(tree, {
          ...defaultOptions,
          name: 'my-app-e2e',
          project: 'my-app',
          bundler: 'vite',
          projectNameAndRootFormat: 'as-provided',
        });

        const cypressConfig = tree.read(
          'my-app-e2e/cypress.config.ts',
          'utf-8'
        );
        expect(cypressConfig).toMatchSnapshot();
      });
    });

    describe('nested', () => {
      it('should set right path names in `cypress.config.ts`', async () => {
        await cypressProjectGenerator(tree, {
          ...defaultOptions,
          name: 'my-app-e2e',
          project: 'my-dir-my-app',
          directory: 'my-dir/my-app-e2e',
          projectNameAndRootFormat: 'as-provided',
        });

        const cypressConfig = tree.read(
          'my-dir/my-app-e2e/cypress.config.ts',
          'utf-8'
        );
        expect(cypressConfig).toMatchSnapshot();
      });

      it('should set right path names in `tsconfig.e2e.json`', async () => {
        await cypressProjectGenerator(tree, {
          ...defaultOptions,
          name: 'my-app-e2e',
          project: 'my-dir-my-app',
          directory: 'my-dir/my-app-e2e',
          projectNameAndRootFormat: 'as-provided',
        });

        const tsconfigJson = readJson(tree, 'my-dir/my-app-e2e/tsconfig.json');
        expect(tsconfigJson).toMatchSnapshot();
      });

      it('should extend from tsconfig.base.json', async () => {
        await cypressProjectGenerator(tree, {
          ...defaultOptions,
          name: 'my-app-e2e',
          project: 'my-app',
          directory: 'my-dir/my-app-e2e',
          projectNameAndRootFormat: 'as-provided',
        });

        const tsConfig = readJson(tree, 'my-dir/my-app-e2e/tsconfig.json');
        expect(tsConfig.extends).toBe('../../tsconfig.base.json');
      });

      it('should support a root tsconfig.json instead of tsconfig.base.json', async () => {
        tree.rename('tsconfig.base.json', 'tsconfig.json');

        await cypressProjectGenerator(tree, {
          ...defaultOptions,
          name: 'my-app-e2e',
          project: 'my-app',
          directory: 'my-dir/my-app-e2e',
          projectNameAndRootFormat: 'as-provided',
        });

        const tsConfig = readJson(tree, 'my-dir/my-app-e2e/tsconfig.json');
        expect(tsConfig.extends).toBe('../../tsconfig.json');
      });

      describe('root project', () => {
        it('should generate in option.name when root project detected', async () => {
          addProjectConfiguration(tree, 'root', { root: '.' });

          await cypressProjectGenerator(tree, {
            ...defaultOptions,
            name: 'e2e-tests',
            baseUrl: 'http://localhost:1234',
            project: 'root',
            projectNameAndRootFormat: 'as-provided',
          });

          expect(tree.listChanges().map((c) => c.path)).toEqual(
            expect.arrayContaining([
              'e2e-tests/cypress.config.ts',
              'e2e-tests/src/e2e/app.cy.ts',
              'e2e-tests/src/fixtures/example.json',
              'e2e-tests/src/support/app.po.ts',
              'e2e-tests/src/support/commands.ts',
              'e2e-tests/src/support/e2e.ts',
              'e2e-tests/tsconfig.json',
            ])
          );
        });

        it('should not generate a root project when the passed in project is not the root project', async () => {
          addProjectConfiguration(tree, 'root', { root: '.' });
          addProjectConfiguration(tree, 'my-cool-app', { root: 'my-cool-app' });

          await cypressProjectGenerator(tree, {
            ...defaultOptions,
            name: 'e2e-tests',
            baseUrl: 'http://localhost:1234',
            project: 'my-cool-app',
            projectNameAndRootFormat: 'as-provided',
          });

          expect(tree.listChanges().map((c) => c.path)).toEqual(
            expect.arrayContaining([
              'e2e-tests/cypress.config.ts',
              'e2e-tests/src/e2e/app.cy.ts',
              'e2e-tests/src/fixtures/example.json',
              'e2e-tests/src/support/app.po.ts',
              'e2e-tests/src/support/commands.ts',
              'e2e-tests/src/support/e2e.ts',
              'e2e-tests/tsconfig.json',
            ])
          );
        });
      });
    });

    describe('--project', () => {
      describe('none', () => {
        it('should not add any implicit dependencies', async () => {
          await cypressProjectGenerator(tree, {
            ...defaultOptions,
            name: 'my-app-e2e',
            baseUrl: 'http://localhost:7788',
            projectNameAndRootFormat: 'as-provided',
          });

          const projectConfig = readProjectConfiguration(tree, 'my-app-e2e');
          expect(projectConfig.implicitDependencies).not.toBeDefined();
          expect(projectConfig.tags).toEqual([]);
        });
      });

      it('should not throw an error when --project does not have targets', async () => {
        const projectConf = readProjectConfiguration(tree, 'my-app');
        delete projectConf.targets;
        updateProjectConfiguration(tree, 'my-app', projectConf);

        await cypressProjectGenerator(tree, {
          name: 'my-app-e2e',
          project: 'my-app',
          linter: Linter.EsLint,
          projectNameAndRootFormat: 'as-provided',
        });

        const projectConfig = readProjectConfiguration(tree, 'my-app-e2e');
        expect(projectConfig.targets['e2e'].options.devServerTarget).toEqual(
          'my-app:serve'
        );
      });
    });

    it('should generate in the correct folder', async () => {
      await cypressProjectGenerator(tree, {
        ...defaultOptions,
        name: 'one-two-other-e2e',
        project: 'my-app',
        directory: 'one/two/other-e2e',
        projectNameAndRootFormat: 'as-provided',
      });

      const project = readProjectConfiguration(tree, 'one-two-other-e2e');
      expect(project).toBeDefined();
      [
        'one/two/other-e2e/cypress.config.ts',
        'one/two/other-e2e/src/e2e/app.cy.ts',
      ].forEach((path) => expect(tree.exists(path)).toBeTruthy());
    });

    it('should generate in the correct folder when --project-name-and-root-format=derived', async () => {
      await cypressProjectGenerator(tree, {
        ...defaultOptions,
        name: 'other-e2e',
        project: 'my-app',
        directory: 'one/two',
        projectNameAndRootFormat: 'derived',
      });

      const project = readProjectConfiguration(tree, 'one-two-other-e2e');
      expect(project).toBeDefined();
      [
        'apps/one/two/other-e2e/cypress.config.ts',
        'apps/one/two/other-e2e/src/e2e/app.cy.ts',
      ].forEach((path) => expect(tree.exists(path)).toBeTruthy());
    });

    describe('serve-static', () => {
      it('should configure Cypress with ci configuration if serve-static is found', async () => {
        const appConfig = readProjectConfiguration(tree, 'my-app');
        appConfig.targets['serve-static'] = {
          executor: 'serve-static-executor',
          options: {},
          configurations: {
            production: {},
          },
        };
        updateProjectConfiguration(tree, 'my-app', appConfig);

        await cypressProjectGenerator(tree, {
          ...defaultOptions,
          name: 'my-app-e2e',
          project: 'my-app',
          projectNameAndRootFormat: 'as-provided',
        });

        const e2eConfig = readProjectConfiguration(tree, 'my-app-e2e');
        expect(e2eConfig.targets.e2e).toMatchObject({
          options: {
            devServerTarget: 'my-app:serve',
          },
          configurations: {
            production: { devServerTarget: 'my-app:serve:production' },
            ci: { devServerTarget: 'my-app:serve-static' },
          },
        });
      });

      it('should not configure Cypress with ci configuration if serve-static is not found', async () => {
        await cypressProjectGenerator(tree, {
          ...defaultOptions,
          name: 'my-app-e2e',
          project: 'my-app',
          projectNameAndRootFormat: 'as-provided',
        });

        const e2eConfig = readProjectConfiguration(tree, 'my-app-e2e');
        expect(e2eConfig.targets.e2e.configurations.ci).toBeUndefined();
      });
    });
  });

  describe('v9 - v7', () => {
    beforeEach(() => {
      mockedInstalledCypressVersion.mockReturnValue(9);
    });

    it('should generate files', async () => {
      await cypressProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
        projectNameAndRootFormat: 'as-provided',
      });

      expect(tree.exists('my-app-e2e/cypress.json')).toBeTruthy();
      expect(tree.exists('my-app-e2e/src/fixtures/example.json')).toBeTruthy();
      expect(
        tree.exists('my-app-e2e/src/integration/app.spec.ts')
      ).toBeTruthy();
      expect(tree.exists('my-app-e2e/src/support/app.po.ts')).toBeTruthy();
      expect(tree.exists('my-app-e2e/src/support/commands.ts')).toBeTruthy();
      expect(tree.exists('my-app-e2e/src/support/index.ts')).toBeTruthy();
    });
  });

  describe('< v7', () => {
    beforeEach(() => {
      mockedInstalledCypressVersion.mockReturnValue(6);
    });

    it('should generate a plugin file if cypress is below version 7', async () => {
      await cypressProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
        projectNameAndRootFormat: 'as-provided',
      });

      expect(tree.exists('my-app-e2e/src/plugins/index.js')).toBeTruthy();
    });

    it('should update project configuration', async () => {
      await cypressProjectGenerator(tree, {
        name: 'my-app-e2e',
        project: 'my-app',
        linter: Linter.EsLint,
        projectNameAndRootFormat: 'as-provided',
      });

      const project = readProjectConfiguration(tree, 'my-app-e2e');
      expect(project.root).toEqual('my-app-e2e');
      expect(project.targets).toMatchSnapshot();
    });

    it('should update project configuration (baseUrl)', async () => {
      await cypressProjectGenerator(tree, {
        name: 'my-app-e2e',
        project: 'my-app',
        baseUrl: 'http://localhost:3000',
        linter: Linter.EsLint,
        projectNameAndRootFormat: 'as-provided',
      });

      const project = readProjectConfiguration(tree, 'my-app-e2e');
      expect(project.root).toEqual('my-app-e2e');
      expect(project.targets).toMatchSnapshot();
    });

    it('should update target configurations', async () => {
      const originalProject = readProjectConfiguration(tree, 'my-app');
      originalProject.targets.serve.defaultConfiguration = 'development';
      originalProject.targets.serve.configurations.development = {};
      updateProjectConfiguration(tree, 'my-app', originalProject);

      await cypressProjectGenerator(tree, {
        name: 'my-app-e2e',
        project: 'my-app',
        linter: Linter.EsLint,
        projectNameAndRootFormat: 'as-provided',
      });

      const project = readProjectConfiguration(tree, 'my-app-e2e');
      expect(project.root).toEqual('my-app-e2e');
      expect(project.targets).toMatchSnapshot();
    });

    describe('nested', () => {
      it('should update configuration', async () => {
        await cypressProjectGenerator(tree, {
          name: 'my-dir-my-app-e2e',
          project: 'my-dir-my-app',
          directory: 'my-dir/my-app-e2e',
          linter: Linter.EsLint,
          projectNameAndRootFormat: 'as-provided',
        });

        const projectConfig = readProjectConfiguration(
          tree,
          'my-dir-my-app-e2e'
        );
        expect(projectConfig).toBeDefined();
        expect(projectConfig.targets).toMatchSnapshot();
      });
    });

    describe('--linter', () => {
      describe('eslint', () => {
        it('should add eslint-plugin-cypress', async () => {
          await cypressProjectGenerator(tree, {
            name: 'my-app-e2e',
            project: 'my-app',
            linter: Linter.EsLint,
            projectNameAndRootFormat: 'as-provided',
          });

          const packageJson = readJson(tree, 'package.json');
          expect(
            packageJson.devDependencies['eslint-plugin-cypress']
          ).toBeTruthy();
          const eslintrcJson = readJson(tree, 'my-app-e2e/.eslintrc.json');
          expect(eslintrcJson).toMatchSnapshot();
        });
      });
    });

    describe('project with directory in its name', () => {
      beforeEach(async () => {
        await cypressProjectGenerator(tree, {
          name: 'my-dir-my-app-e2e',
          project: 'my-dir-my-app',
          directory: 'my-dir/my-app-e2e',
          linter: Linter.EsLint,
          projectNameAndRootFormat: 'as-provided',
        });
      });

      it('should update configuration', async () => {
        const projectConfig = readProjectConfiguration(
          tree,
          'my-dir-my-app-e2e'
        );
        expect(projectConfig).toBeDefined();
        expect(projectConfig.targets).toMatchSnapshot();
      });

      it('should update nx.json', async () => {
        const project = readProjectConfiguration(tree, 'my-dir-my-app-e2e');
        expect(project.tags).toEqual([]);
        expect(project.implicitDependencies).toEqual(['my-dir-my-app']);
      });

      it('should set right path names in `cypress.json`', async () => {
        const cypressConfig = tree.read(
          'my-dir/my-app-e2e/cypress.json',
          'utf-8'
        );
        expect(cypressConfig).toMatchSnapshot();
      });
    });
  });
});
