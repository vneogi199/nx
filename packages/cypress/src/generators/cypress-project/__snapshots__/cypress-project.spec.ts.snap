// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cypress Project < v7 --linter eslint should add eslint-plugin-cypress 1`] = `
{
  "extends": [
    "plugin:cypress/recommended",
    "../.eslintrc.json",
  ],
  "ignorePatterns": [
    "!**/*",
  ],
  "overrides": [
    {
      "files": [
        "*.ts",
        "*.tsx",
        "*.js",
        "*.jsx",
      ],
      "rules": {},
    },
    {
      "files": [
        "src/plugins/index.js",
      ],
      "rules": {
        "@typescript-eslint/no-var-requires": "off",
        "no-undef": "off",
      },
    },
  ],
}
`;

exports[`Cypress Project < v7 nested should update configuration 1`] = `
{
  "e2e": {
    "configurations": {
      "production": {
        "devServerTarget": "my-dir-my-app:serve:production",
      },
    },
    "executor": "@nx/cypress:cypress",
    "options": {
      "cypressConfig": "my-dir/my-app-e2e/cypress.json",
      "devServerTarget": "my-dir-my-app:serve",
      "testingType": "e2e",
      "tsConfig": "my-dir/my-app-e2e/tsconfig.json",
    },
  },
  "lint": {
    "executor": "@nx/eslint:lint",
    "options": {
      "lintFilePatterns": [
        "my-dir/my-app-e2e/**/*.{js,ts}",
      ],
    },
    "outputs": [
      "{options.outputFile}",
    ],
  },
}
`;

exports[`Cypress Project < v7 project with directory in its name should set right path names in \`cypress.json\` 1`] = `
"{
  "fileServerFolder": ".",
  "fixturesFolder": "./src/fixtures",
  "integrationFolder": "./src/integration",
  "modifyObstructiveCode": false,
  "supportFile": "./src/support/index.ts",
  "pluginsFile": "./src/plugins/index",
  "video": true,
  "videosFolder": "../../dist/cypress/my-dir/my-app-e2e/videos",
  "screenshotsFolder": "../../dist/cypress/my-dir/my-app-e2e/screenshots",
  "chromeWebSecurity": false
}
"
`;

exports[`Cypress Project < v7 project with directory in its name should update configuration 1`] = `
{
  "e2e": {
    "configurations": {
      "production": {
        "devServerTarget": "my-dir-my-app:serve:production",
      },
    },
    "executor": "@nx/cypress:cypress",
    "options": {
      "cypressConfig": "my-dir/my-app-e2e/cypress.json",
      "devServerTarget": "my-dir-my-app:serve",
      "testingType": "e2e",
      "tsConfig": "my-dir/my-app-e2e/tsconfig.json",
    },
  },
  "lint": {
    "executor": "@nx/eslint:lint",
    "options": {
      "lintFilePatterns": [
        "my-dir/my-app-e2e/**/*.{js,ts}",
      ],
    },
    "outputs": [
      "{options.outputFile}",
    ],
  },
}
`;

exports[`Cypress Project < v7 should update project configuration (baseUrl) 1`] = `
{
  "e2e": {
    "executor": "@nx/cypress:cypress",
    "options": {
      "baseUrl": "http://localhost:3000",
      "cypressConfig": "my-app-e2e/cypress.json",
      "testingType": "e2e",
      "tsConfig": "my-app-e2e/tsconfig.json",
    },
  },
  "lint": {
    "executor": "@nx/eslint:lint",
    "options": {
      "lintFilePatterns": [
        "my-app-e2e/**/*.{js,ts}",
      ],
    },
    "outputs": [
      "{options.outputFile}",
    ],
  },
}
`;

exports[`Cypress Project < v7 should update project configuration 1`] = `
{
  "e2e": {
    "configurations": {
      "production": {
        "devServerTarget": "my-app:serve:production",
      },
    },
    "executor": "@nx/cypress:cypress",
    "options": {
      "cypressConfig": "my-app-e2e/cypress.json",
      "devServerTarget": "my-app:serve",
      "testingType": "e2e",
      "tsConfig": "my-app-e2e/tsconfig.json",
    },
  },
  "lint": {
    "executor": "@nx/eslint:lint",
    "options": {
      "lintFilePatterns": [
        "my-app-e2e/**/*.{js,ts}",
      ],
    },
    "outputs": [
      "{options.outputFile}",
    ],
  },
}
`;

exports[`Cypress Project < v7 should update target configurations 1`] = `
{
  "e2e": {
    "configurations": {
      "production": {
        "devServerTarget": "my-app:serve:production",
      },
    },
    "executor": "@nx/cypress:cypress",
    "options": {
      "cypressConfig": "my-app-e2e/cypress.json",
      "devServerTarget": "my-app:serve:development",
      "testingType": "e2e",
      "tsConfig": "my-app-e2e/tsconfig.json",
    },
  },
  "lint": {
    "executor": "@nx/eslint:lint",
    "options": {
      "lintFilePatterns": [
        "my-app-e2e/**/*.{js,ts}",
      ],
    },
    "outputs": [
      "{options.outputFile}",
    ],
  },
}
`;

exports[`Cypress Project > v10 for bundler:vite should pass the bundler info to nxE2EPreset in \`cypress.config.ts\` 1`] = `
"import { defineConfig } from 'cypress';
import { nxE2EPreset } from '@nx/cypress/plugins/cypress-preset';

export default defineConfig({
  e2e: nxE2EPreset(__dirname, {
    bundler: 'vite',
  }),
});
"
`;

exports[`Cypress Project > v10 nested should set right path names in \`cypress.config.ts\` 1`] = `
"import { defineConfig } from 'cypress';
import { nxE2EPreset } from '@nx/cypress/plugins/cypress-preset';

export default defineConfig({
  e2e: nxE2EPreset(__dirname),
});
"
`;

exports[`Cypress Project > v10 nested should set right path names in \`tsconfig.e2e.json\` 1`] = `
{
  "compilerOptions": {
    "allowJs": true,
    "outDir": "../../dist/out-tsc",
    "sourceMap": false,
    "types": [
      "cypress",
      "node",
    ],
  },
  "extends": "../../tsconfig.base.json",
  "include": [
    "src/**/*.ts",
    "src/**/*.js",
    "cypress.config.ts",
  ],
}
`;

exports[`Cypress Project > v10 should set right path names in \`cypress.config.ts\` 1`] = `
"import { defineConfig } from 'cypress';
import { nxE2EPreset } from '@nx/cypress/plugins/cypress-preset';

export default defineConfig({
  e2e: nxE2EPreset(__dirname),
});
"
`;

exports[`Cypress Project > v10 should set right path names in \`tsconfig.e2e.json\` 1`] = `
{
  "compilerOptions": {
    "allowJs": true,
    "outDir": "../dist/out-tsc",
    "sourceMap": false,
    "types": [
      "cypress",
      "node",
    ],
  },
  "extends": "../tsconfig.base.json",
  "include": [
    "src/**/*.ts",
    "src/**/*.js",
    "cypress.config.ts",
  ],
}
`;

exports[`Cypress Project > v10 should update configuration when eslint is passed 1`] = `
{
  "executor": "@nx/eslint:lint",
  "options": {
    "lintFilePatterns": [
      "my-app-e2e/**/*.{js,ts}",
    ],
  },
  "outputs": [
    "{options.outputFile}",
  ],
}
`;
